### PR TITLE
Package dapp-tests with Nix

### PR DESCRIPTION
**Pull request by asymmetric**
@MrChico this should be an alternative to using submodules and would close #404.

Draft because I need to clean up the history.

The main issue I had was that `solidityPackage` uses its own bulder, which means it was hard to adapt it to our needs - which are a bit different than usual, since we not only want to test by running `hevm dapp-test` on the contracts but also want to run a bash script.

So I've moved most of its logic to a `buildDappPackage` function, which uses the usual `mkDerivation` infra and doesn't require its own builder, and allows for easier overriding and adding of phases.

Ideally we would remove/alias `solidityPackage`, but I want to make sure it works for all consumers (cc @icetan).

### CODE REVIEW

File: nix/build-dapp-package.nix [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+      REMAPPINGS =
+        pkgs.lib.mapAttrsToList
+          (k: v: k + "=" + v)
+          passthru.remappings;
+
+      LIBSCRIPT =
+        pkgs.lib.mapAttrsToList
+          (k: v: ''
+            ln -s ${v} lib/${k}
```
Comment (asymmetric) [inline nix/build-dapp-package.nix:45]: I think we could probably use `symlinkJoin` instead of this.

File: src/dapp-tests/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{ pkgs }:
+
+let
+  ds-test-src = pkgs.fetchFromGitHub {
+    owner = "dapphub";
+    repo = "ds-test";
+    rev = "eb7148d43c1ca6f9890361e2e2378364af2430ba";
```
Comment (asymmetric) [inline src/dapp-tests/default.nix:7]: Very open to using niv here.

File: overlay.nix [hunk 3]
```diff
@@ -133,6 +118,8 @@ in rec {
   seth = self.callPackage (import ./src/seth) {};
   dapp = self.callPackage (import ./src/dapp) {};
 
+  dapp-tests = self.callPackage (import ./src/dapp-tests) {};
```
Comment (MrChico) [inline overlay.nix:121]: I thought you wanted to remove this from the overlay?

Comment (asymmetric) [reply overlay.nix:121]: That was when we thought it would just be a shell script.

But because these integration tests are both dapp packages and other stuff (like a shell script), then I think it makes sense to use our solidity -> nix infrastructure, run tests with hevm and run anything else we want to run.

File: src/dapp-tests/default.nix (hunk 2 above)
Comment (asymmetric) [reply src/dapp-tests/default.nix:7]: But maybe in a separate PR, where we use it for the whole repo and not just for this library.

File: nix/build-dapp-package.nix [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        export DAPP_LIB=lib
+      '';
+
+      checkPhase = args.checkPhase or ''
+        hevm dapp-test
+      '';
+      installPhase = args.installPhase or ''
+        mkdir -p $out/dapp/$name
+        cp -r $src $out/dapp/$name/src
+        cp -r lib $out/dapp/$name/lib
```
Comment (asymmetric) [inline nix/build-dapp-package.nix:72]: It could be interesting to do a multiple outputs derivation instead of this, so that they would result in different store paths.

This could mean that if e.g. a dependency doesn't change, that store path remains unchanged.

File: overlay.nix [hunk 5]
```diff
@@ -101,19 +86,19 @@ in rec {
 
   solc-versions =
     let
-      fetchNixpkgs = { owner, attr }:
+      fetchSolcVersions = { owner, attr }:
```
Comment (asymmetric) [inline overlay.nix:89]: These changes I've just ninja'd in, they have no relationship to this PR :)

File: nix/build-dapp-package.nix [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+      LIBSCRIPT =
+        pkgs.lib.mapAttrsToList
+          (k: v: ''
+            ln -s ${v} lib/${k}
+          '')
+          passthru.libPaths;
+
+      buildPhase = args.buildPhase or ''
+        jsonopts=--combined-json=abi,bin,bin-runtime,srcmap,srcmap-runtime,ast
```
Comment (asymmetric) [inline nix/build-dapp-package.nix:50]: NOTE: I noticed this is not taking into account the recent changes introduced in #400, fixing this.

File: nix/build-dapp-package.nix [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        (x: {
+          "${x.name}" = "${x}/dapp/${x.name}/src";
+         } // x.libPaths)
+         xs);
+in
+  {
+    name
+  , src
+  , deps ? []
+  , extract ? true
```
Comment (MrChico) [inline nix/build-dapp-package.nix:23]: does this mean that it defaults to true? I don't think `extract` is commonly used

File: overlay.nix (hunk 5 above)
Comment (MrChico) [reply overlay.nix:93]: is it just name changes or does this change the semantics?

Comment (asymmetric) [reply overlay.nix:93]: Only name

File: nix/build-dapp-package.nix (hunk 7 above)
Comment (asymmetric) [reply nix/build-dapp-package.nix:23]: Right. I took it from [here](https://github.com/dapphub/dapptools/pull/400/files#diff-a8c475bf1172056e5043d77688ed2e1bR20), but it makes sense that it should default to `false`, probably.

File: nix/build-dapp-package.nix (hunk 4 above)
Comment (MrChico) [reply nix/build-dapp-package.nix:85]: deduplication?

File: src/dapp-tests/default.nix [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    doCheck = false;
+  };
+
+in
+  pkgs.buildDappPackage {
+    name = "dapp-tests";
+    src = ./.;
+    deps = [ ds-test ];
+
+    solcFlags = "--evm-version constantinople";
```
Comment (MrChico) [inline src/dapp-tests/default.nix:23]: I think it's better to leave the default `evm-version` instead. The current version of solidity should default to Istanbul, methinks

File: src/dapp-tests/default.nix [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+let
+  ds-test-src = pkgs.fetchFromGitHub {
+    owner = "dapphub";
+    repo = "ds-test";
+    rev = "eb7148d43c1ca6f9890361e2e2378364af2430ba";
+    sha256 = "1phnqjkbcqg18mh62c8jq0v8fcwxs8yc4sa6dca4y8pq2k35938k";
+  } + "/src";
+
+  ds-test = pkgs.buildDappPackage {
```
Comment (MrChico) [inline src/dapp-tests/default.nix:11]: cool!

File: nix/build-dapp-package.nix [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            echo "$contract" | jq '.["abi"]' -r > "out/$fileName.abi"
+            echo "$contract" | jq '.["bin"]' -r > "out/$fileName.bin"
+            echo "$contract" | jq '.["bin-runtime"]' -r > "out/$fileName.bin-runtime"
+            echo "$contract" | jq '.["metadata"]' -r > "out/$fileName.metadata"
+          done
+        fi
+      '';
+
+      checkPhase = args.checkPhase or ''
+        DAPP_OUT=out dapp2-test-hevm
```
Comment (MrChico) [inline nix/build-dapp-package.nix:79]: out is the default value for `DAPP_OUT` FYI

File: src/dapp-tests/default.nix (hunk 2 above)
Comment (MrChico) [reply src/dapp-tests/default.nix:7]: I could see us using `master` here, to ensure that what's being tested is the same as people's experience using `dapp install ds-test`

File: src/seth/default.nix [hunk 11]
```diff
@@ -6,10 +6,11 @@ stdenv.mkDerivation rec {
   version = "0.9.0";
   src = ./.;
 
+  buildInputs = [ nodejs ];
```
Comment (MrChico) [inline src/seth/default.nix:9]: why is this needed? `nodejs` is already added to path in `postinstall`?

File: src/seth/default.nix [hunk 12]
```diff
@@ -6,10 +6,11 @@ stdenv.mkDerivation rec {
...
   nativeBuildInputs = [makeWrapper shellcheck];
-  buildPhase = "true";
+  dontBuild = true;
```
Comment (MrChico) [inline src/seth/default.nix:11]: the linking step is not needed? Its being done with `patchShebangs` or?

File: nix/build-dapp-package.nix [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+          passthru.remappings;
+
+      LIBSCRIPT =
+        pkgs.lib.mapAttrsToList
+          (k: v: ''
+            ln -s ${v} lib/${k}
+          '')
+          passthru.libPaths;
+
+      buildPhase = args.buildPhase or ''
```
Comment (MrChico) [inline nix/build-dapp-package.nix:50]: Is there any way we could merge this and `dapp build` somehow? This script is essentially `dapp build`. It seems that whenever we change `dapp build` we'd also need to change it here.

File: nix/build-dapp-package.nix (hunk 10 above)
Comment (asymmetric) [reply nix/build-dapp-package.nix:79]: Where do you see that? The default would have to be set [here](
https://github.com/dapphub/dapptools/tree/15c241a82183da46f5e170350dd378c99e1f264b/nix/dapp/dapp-test-hevm.nix#L5) for it to take effect.

File: src/seth/default.nix (hunk 12 above)
Comment (asymmetric) [reply src/seth/default.nix:11]: seth doesn't need linking. At runtime, it calls a file named `seth-$subcommand` in libexec, see [here](
https://github.com/dapphub/dapptools/tree/15c241a82183da46f5e170350dd378c99e1f264b/src/seth/libexec/seth/seth#L174)

File: src/seth/default.nix (hunk 11 above)
Comment (asymmetric) [reply src/seth/default.nix:9]: This is needed so that `patchShebangs` finds `noejs` in the `PATH` and therefore patches it.

I'll see if it's possible to remove the repetition, but it could be weird if only `nodejs` is treated differently compared to all the other packages in the `path` variable below.

File: src/dapp-tests/default.nix (hunk 9 above)
Comment (asymmetric) [reply src/dapp-tests/default.nix:11]: Right?? :)

File: nix/build-dapp-package.nix (hunk 4 above)
Comment (asymmetric) [reply nix/build-dapp-package.nix:85]: Right, maybe for `src` we would get some deduplication in some (rare?) cases. Not for `lib` though, there are only symlinks in there.

I think it's pretty cool aesthetically so we should probably do it anyway

File: nix/build-dapp-package.nix (hunk 10 above)
Comment (asymmetric) [reply nix/build-dapp-package.nix:79]: dapp2-test-hevm is not in `libexec/`, meaning the `libexec/dapp` script is not run, so variables exported there have no effect.

File: src/dapp-tests/default.nix (hunk 2 above)
Comment (asymmetric) [reply src/dapp-tests/default.nix:7]: WDYM with using master? I took the last commit on master on the repo, and we have to pin one commit for this to be reproducible.. What do you have in mind?

File: src/seth/default.nix (hunk 12 above)
Comment (asymmetric) [reply src/seth/default.nix:11]: Or what linking step are you referring to?

File: nix/build-dapp-package.nix (hunk 13 above)
Comment (asymmetric) [reply nix/build-dapp-package.nix:50]: Hmm yeah, that's a good point.

Maybe we could extract the bash stuff in the different phases into their own little script, and use those both in `build-dapp-package` and in `dapp-build`.

Is it OK if I file an issue for this or do you think it should be in this PR?

Comment (asymmetric) [reply nix/build-dapp-package.nix:50]: Also honestly, inlined Bash inside of Nix is pretty gross. It would be good to move it out for that reason too.

File: nix/build-dapp-package.nix (hunk 10 above)
Comment (MrChico) [reply nix/build-dapp-package.nix:79]: Im wrobg

File: src/seth/default.nix (hunk 12 above)
Comment (MrChico) [reply src/seth/default.nix:11]: I'm referring to what was hapenning in `make install`

File: nix/build-dapp-package.nix (hunk 13 above)
Comment (MrChico) [reply nix/build-dapp-package.nix:50]: Yeah, we can def postpone this for a later pr

**Comment by MrChico** [comment]
This lgtm. Commits will be squashed when we merge anyway

File: nix/build-dapp-package.nix (hunk 10 above)
Comment (asymmetric) [reply nix/build-dapp-package.nix:79]: What??

**Comment by asymmetric** [comment]
No let’s not squash them please, it would make the PR too over the place IMO.

**Comment by icetan** [comment]
@asymmetric very cool, I see no reason for this not replacing `solidityPackage`.

File: nix/build-dapp-package.nix (hunk 10 above)
Comment (MrChico) [reply nix/build-dapp-package.nix:79]: I retract my previous statement

File: src/dapp-tests/default.nix (hunk 2 above)
Comment (MrChico) [reply src/dapp-tests/default.nix:7]: I had in mind not pinning it, making it non reproducible with the idea that whats being tested here is the same thing that people will be using

File: src/seth/default.nix (hunk 11 above)
Comment (asymmetric) [reply src/seth/default.nix:9]: But the reason why patching the `#!/usr/bin/env node` shebang is necessary for seth used in `dapp-tests` (and isn't necessary for seth used from the shell to work) is not clear to me.

File: src/dapp-tests/default.nix (hunk 2 above)
Comment (asymmetric) [reply src/dapp-tests/default.nix:7]: You can't make it non-repro, AFAIK.

File: src/seth/default.nix (hunk 12 above)
Comment (asymmetric) [reply src/seth/default.nix:11]: The `buildPhase` by default [just runs `make`](https://nixos.org/nixpkgs/manual/#build-phase), which in our case is `test` - so it was running tests, which I found weird.

Comment (asymmetric) [reply src/seth/default.nix:11]: The `buildPhase` above was just running `true` btw, as in the program `true`.
